frontends/p4/validateParsedProgram.cpp

```diff
@@ -225,7 +225,8 @@ void ValidateParsedProgram::postorder(const IR::P4Program* program) {
         cstring name = decl->getName();
         auto existing = declarations.getDeclaration(name);
         if (existing != nullptr) {
-            if (!existing->is<IR::IFunctional>() || !decl->is<IR::IFunctional>()) {
+            if (!existing->is<IR::IFunctional>() || !decl->is<IR::IFunctional>()
+                || typeid(*existing) != typeid(*decl) || decl->is<IR::P4Action>()) {
```

Comment:
I'm surprised that there is a `decl->is<IR::P4Action>()`, but not a matching `existing->is<IR::P4Action>()`. Is that missing here?

Reply:
I think that the typeid will take care of that.

Reply:
I guess not, it's an or.

Reply:
yes, but it's a negated or. so it should work

Reply:
Ah, I see. Because of the `typeid` check, we'd expect `decl` and `existing` to have the same type, and so, adding an `|| existing->is<IR::P4Action>()` would be redundant.